src/transaction/log_replication_atomic.cpp

```diff
@@ -402,41 +408,122 @@ namespace cublog
 	assert_release (false);
       }
 
-#if !defined (NDEBUG)
-    /* there will be no duplicated object in the map */
-    auto [begin, end] = m_locked_classes.equal_range (trid);
-    if (std::count_if (begin, end, [classoid] (const auto &it)
-    {
-      return it.second == *classoid;
-    }) > 0)
-    {
-      assert (false);
-    }
-#endif
+    m_locked_classes[trid].emplace (*classoid);
+  }
 
-    m_locked_classes.emplace (trid, *classoid);
+  void
+  atomic_replicator::bookkeep_classname (cubthread::entry &thread_entry, const OID *classoid)
+  {
+    char *classname = nullptr;
+
+    assert (!OID_ISNULL (classoid) && !OID_ISTEMP (classoid));
+
+    (void) heap_get_class_name (&thread_entry, classoid, &classname);
```

Comment:
No need to check the returned error?

Reply:
hmm.. I can add an assert for the error code.
when CREATE TABLE is replicated, 
`classname == nullptr` and error code will be set to `ER_HEAP_UNKNOWN_OBJECT`.

How about adding an assertion when 
if classname == nullptr, then error ==  ER_HEAP_UNKNOWN_OBJECT
else classname != nullptr, then error ==  ER_HEAP_UNKNOWN_OBJECT

then we can filter out other failure cases (e.g ER_FAILED)

Reply:
OK. I think this way is enough, 
```
if (error != NO_ERROR)
{
   assert (error == ER_HEAP_UNKNOWN_OBJECT);
}
```

Reply:
How about branching based on the error code, not by the `classname`? It seems better since the return value means "what's happened" and intuitive. like
```
if (error == NO_ERROR)
{
 assert (class_name != nullptr);
  m_classname_map.emplace (*classoid, std::move (std::string (classname)));
} 
else if (error == ER_HEAP_UNKNOWN_OBJECT)
{
  assert (class_name == nullptr);
  /* no class record means it's creating the class */
}
else 
{
  assert (false);
}
```

Reply:
The same goes for the `update_classname_cache_for_ddl`.

Reply:
I agree, I will modify as you suggest